### ISSUE TITLE
[REQ] Skip request to container registry when image already exists locally

### ISSUE THREAD

**Issue by damienpontifex** [project-copacetic/copacetic#177]
### What kind of request is this?

Improvement of existing experience

### What is your request or suggestion?

I would like copa to be able to scan an image built within a pipeline that is tagged with a remote hostname, but exists locally.

In our setup we build an image and tag `registry-name.azurecr.io/repo:tag`. At this point in time (build only / PR validation) the pipeline isn't authenticated to the container registry. Trivy scans this after build using the local image. We've then setup to run copa on this image but currently it fails with the debug log:

```log
time="2023-06-09T06:17:02Z" level=debug msg=resolving host=registry-name.azurecr.io
time="2023-06-09T06:17:02Z" level=debug msg="do request" host=registry-name.azurecr.io request.header.accept="application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, */*" request.header.user-agent=buildkit/0.0.0+unknown request.method=HEAD url="https://registry-name.azurecr.io/v2/repo/manifests/tag"
time="2023-06-09T06:17:02Z" level=debug msg="fetch response received" host=registry-name.azurecr.io response.header.access-control-expose-headers=Docker-Content-Digest response.header.access-control-expose-headers.1=WWW-Authenticate response.header.access-control-expose-headers.2=Link response.header.access-control-expose-headers.3=X-Ms-Correlation-Request-Id response.header.connection=keep-alive response.header.content-length=211 response.header.content-type="application/json; charset=utf-8" response.header.date="Fri, 09 Jun 2023 06:17:02 GMT" response.header.docker-distribution-api-version=registry/2.0 response.header.server=openresty response.header.strict-transport-security="max-age=31536000; includeSubDomains" response.header.strict-transport-security.1="max-age=31536000; includeSubDomains" response.header.www-authenticate="Bearer realm=\"https://registry-name.azurecr.io/oauth2/token\",service=\"registry-name.azurecr.io\",scope=\"repository:repo:pull\"" response.header.x-content-type-options=nosniff response.header.x-ms-correlation-request-id=60b02c26-0626-4ffe-b00e-4a39acb5ccb7 response.status="401 Unauthorized" url="https://registry-name.azurecr.io/v2/repo/manifests/tag"
time="2023-06-09T06:17:02Z" level=debug msg=Unauthorized header="Bearer realm=\"https://registry-name.azurecr.io/oauth2/token\",service=\"registry-name.azurecr.io\",scope=\"repository:repo:pull\"" host=registry-name.azurecr.io
time="2023-06-09T06:17:02Z" level=debug msg="do request" host=registry-name.azurecr.io request.header.accept="application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, */*" request.header.user-agent=buildkit/0.0.0+unknown request.method=HEAD url="https://registry-name.azurecr.io/v2/repo/manifests/tag"
time="2023-06-09T06:17:02Z" level=info msg="trying next host" error="failed to authorize: failed to fetch anonymous token: unexpected status from GET request to https://registry-name.azurecr.io/oauth2/token?scope=repository%3Arepo%3Apull&service=registry-name.azurecr.io: 401 Unauthorized" host=registry-name.azurecr.io
time="2023-06-09T06:17:02Z" level=warning msg="--debug specified, working folder at /tmp/copa-1306459613 needs to be manually cleaned up"
Error: failed to authorize: failed to fetch anonymous token: unexpected status from GET request to https://registry-name.azurecr.io/oauth2/token?scope=repository%3Arepo%3Apull&service=registry-name.azurecr.io: 401 Unauthorized
##[debug]Exit code 1 received from tool '/usr/bin/bash'
```

If the image already exists on the build machine, can this request and hence authentication process be skipped and work with the local image?

**Comment by sozercan** [comment]
Related #196

**Comment by pmengelbert** [comment]
@damienpontifex I'd like to tackle this issue. Could you post steps to reproduce? It'll help move things along quickly

**Comment by sozercan** [comment]
@pmengelbert here is an example

```
docker pull ubuntu:focal-20210827
docker tag local/ubuntu:focal-20210827
trivy ...
copa patch ...
```

**Comment by pmengelbert** [comment]
@damienpontifex This has been implemented and merged in #381 . Please see the docs [here](https://github.com/project-copacetic/copacetic/blob/e5288e885d594bae3ba2f21bdf176ea9c6fb87de/website/docs/quick-start.md?plain=1#L11C1-L16C111) and [here](https://github.com/project-copacetic/copacetic/blob/e5288e885d594bae3ba2f21bdf176ea9c6fb87de/website/docs/quick-start.md?plain=1#L104C1-L110C72) for the limitations of this feature.

**Comment by damienpontifex** [comment]
Thanks. We'll try it out on our pipelines
